crates/context/config/src/client.rs

```diff
@@ -0,0 +1,323 @@
+        Self { transport }
+    }
+}
+
+impl<T: Transport> ContextConfigClient<T> {
+    pub fn query<'a>(
+        &'a self,
+        network_id: Cow<'a, str>,
+        contract_id: Cow<'a, str>,
+    ) -> ContextConfigQueryClient<'a, T> {
+        ContextConfigQueryClient {
+            network_id,
+            contract_id,
+            transport: &self.transport,
+        }
+    }
+
+    pub fn mutate<'a>(
+        &'a self,
+        network_id: Cow<'a, str>,
+        contract_id: Cow<'a, str>,
+        signer_id: SignerId,
+    ) -> ContextConfigMutateClient<'a, T> {
+        ContextConfigMutateClient {
+            network_id,
+            contract_id,
+            signer_id,
+            transport: &self.transport,
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum Error<T: Transport> {
+    #[error("transport error: {0}")]
+    Transport(T::Error),
+    #[error(transparent)]
+    Other(#[from] types::Error<Infallible>),
+}
+
+#[derive(Debug)]
+pub struct Response<T> {
+    bytes: Vec<u8>,
+    _priv: std::marker::PhantomData<T>,
+}
+
+impl<T> Response<T> {
+    fn new(bytes: Vec<u8>) -> Self {
+        Self {
+            bytes,
+            _priv: Default::default(),
+        }
+    }
+
+    pub fn parse(&self) -> Result<T, serde_json::Error>
+    where
+        T: for<'a> Deserialize<'a>,
+    {
+        serde_json::from_slice(&self.bytes)
+    }
+}
+
+#[derive(Debug)]
+pub struct ContextConfigQueryClient<'a, T> {
+    network_id: Cow<'a, str>,
+    contract_id: Cow<'a, str>,
+    transport: &'a T,
+}
+
+impl<'a, T: Transport> ContextConfigQueryClient<'a, T> {
+    async fn read<I: Serialize, O>(&self, method: &str, body: I) -> Result<Response<O>, Error<T>> {
+        let payload = serde_json::to_vec(&body).map_err(|err| Error::Other(err.into()))?;
+
+        let request = TransportRequest {
+            network_id: Cow::Borrowed(&self.network_id),
+            contract_id: Cow::Borrowed(&self.contract_id),
+            operation: Operation::Read {
+                method: Cow::Borrowed(method),
+            },
+        };
+
+        let response = self
+            .transport
+            .send(request, payload)
+            .await
+            .map_err(Error::Transport)?;
+
+        Ok(Response::new(response))
+    }
+
+    pub async fn application(
+        &self,
+        context_id: ContextId,
+    ) -> Result<Response<Application<'static>>, Error<T>> {
+        self.read(
+            "application",
+            json!({
+                "context_id": Repr::new(context_id),
+            }),
+        )
+        .await
+    }
+
+    pub async fn members(
+        &self,
+        context_id: ContextId,
+        offset: usize,
+        length: usize,
+    ) -> Result<Response<Vec<Repr<ContextIdentity>>>, Error<T>> {
+        self.read(
+            "members",
+            json!({
+                "context_id": Repr::new(context_id),
+                "offset": offset,
+                "length": length,
+            }),
```

Comment:
I am curious... it feels like, in places like this, we would benefit from more formalisation. I.e. using structs, which are then converted to JSON as and when needed. That would, I think, be best/standard practice, and allow us to rely upon Rust's type system...? However, I'm sure there is a good reason why it's like this, and I likely don't know the whole context! 😄 Thoughts?

Reply:
Unlike the accommodations made for `mutate` and `set`, these query methods are transparently what `near_sdk`'s codegen produces, which maintains the structures these are deserialized to privately.

We could introduce a struct, but it would be single-use and private, at this point, there's no added benefit, with `json!` being more straightforward.